dask/dataframe/tests/test_dataframe.py

```diff
@@ -1426,13 +1442,92 @@ def test_concat2():
+    pdf1 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCDE'), index=list('abcdef'))
+    pdf2 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCFG'), index=list('ghijkl'))
+    pdf3 = pd.DataFrame(np.random.randn(6, 5),
+                        columns=list('ABCHI'), index=list('mnopqr'))
+    ddf1 = dd.from_pandas(pdf1, 2)
+    ddf2 = dd.from_pandas(pdf2, 3)
+    ddf3 = dd.from_pandas(pdf3, 2)
+
+    result = dd.concat([ddf1, ddf2])
+    assert result.divisions == ddf1.divisions[:-1] + ddf2.divisions
+    assert result.npartitions == ddf1.npartitions + ddf2.npartitions
+    assert eq(result, pd.concat([pdf1, pdf2]))
+
+    assert eq(dd.concat([ddf1, ddf2], respect_partitions=True),
+              pd.concat([pdf1, pdf2]))
+
+    result = dd.concat([ddf1, ddf2, ddf3])
+    assert result.divisions == (ddf1.divisions[:-1] + ddf2.divisions[:-1] +
+                                ddf3.divisions)
+    assert result.npartitions == (ddf1.npartitions + ddf2.npartitions +
+                                  ddf3.npartitions)
+    assert eq(result, pd.concat([pdf1, pdf2, pdf3]))
+
+    assert eq(dd.concat([ddf1, ddf2, ddf3], respect_partitions=True),
+              pd.concat([pdf1, pdf2, pdf3]))
+
+def test_concat4_respect_partitions():
+    pdf1 = pd.DataFrame(np.random.randn(10, 5),
+                        columns=list('ABCDE'), index=list('abcdefghij'))
+    pdf2 = pd.DataFrame(np.random.randn(13, 5),
+                        columns=list('ABCDE'), index=list('fghijklmnopqr'))
+    pdf3 = pd.DataFrame(np.random.randn(13, 6),
+                        columns=list('CDEXYZ'), index=list('fghijklmnopqr'))
+
+    ddf1 = dd.from_pandas(pdf1, 2)
+    ddf2 = dd.from_pandas(pdf2, 3)
+    ddf3 = dd.from_pandas(pdf3, 2)
+
+    assert raises(ValueError, lambda: dd.concat([ddf1, ddf1]))
+    assert eq(dd.concat([ddf1, ddf1], respect_partitions=True),
+              pd.concat([pdf1, pdf1]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf1, ddf2]))
+    assert eq(dd.concat([ddf1, ddf2], respect_partitions=True),
+              pd.concat([pdf1, pdf2]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf1, ddf3]))
+    assert eq(dd.concat([ddf1, ddf3], respect_partitions=True),
+              pd.concat([pdf1, pdf3]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf2, ddf1]))
+    assert eq(dd.concat([ddf2, ddf1], respect_partitions=True),
+              pd.concat([pdf2, pdf1]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf2, ddf3]))
+    assert eq(dd.concat([ddf2, ddf3], respect_partitions=True),
+              pd.concat([pdf2, pdf3]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf3, ddf1]))
+    assert eq(dd.concat([ddf3, ddf1], respect_partitions=True),
+              pd.concat([pdf3, pdf1]))
+
+    assert raises(ValueError, lambda: dd.concat([ddf3, ddf2]))
+    assert eq(dd.concat([ddf3, ddf2], respect_partitions=True),
+              pd.concat([pdf3, pdf2]))
```

Comment:
Yeah, I think it's fine to add new keyword arguments judiciously when appropriate, as long as the default behavior matches pandas. So this looks like the right solution to me.